crates/consensus/beacon/src/engine/mod.rs

```diff
@@ -0,0 +1,387 @@
+use futures::{Future, FutureExt, StreamExt};
+use reth_db::{database::Database, tables, transaction::DbTx};
+use reth_executor::blockchain_tree::{BlockStatus, BlockchainTree};
+use reth_interfaces::{
+    consensus::{Consensus, ForkchoiceState},
+    sync::SyncStateUpdater,
+};
+use reth_primitives::{BlockHash, SealedBlock, H256};
+use reth_provider::ExecutorFactory;
+use reth_rpc_types::engine::PayloadStatusEnum;
+use reth_stages::Pipeline;
+use std::{
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tokio::sync::mpsc::UnboundedReceiver;
+use tokio_stream::wrappers::UnboundedReceiverStream;
+use tracing::*;
+
+mod error;
+pub use error::BeaconEngineError;
+
+mod message;
+pub use message::BeaconEngineMessage;
+
+mod pipeline_state;
+pub use pipeline_state::PipelineState;
+
+#[derive(Debug, Default)]
+enum BeaconEngineAction {
+    #[default]
+    None,
+    RunPipeline,
+}
+
+impl BeaconEngineAction {
+    fn run_pipeline(&self) -> bool {
+        matches!(self, BeaconEngineAction::RunPipeline)
+    }
+}
+
+/// The beacon consensus engine is the driver that switches between historical and live sync.
+///
+/// TODO: add more docs
```

Comment:
Just adding a comment here to highlight so it is visible that this needs to be done.

What I would like to be added at least is the distinction between historical and live sync (e.g. that one uses the pipeline which is faster because xyz and so on)